packages/diagrams/lib/src/animated_icons.dart

```diff
@@ -0,0 +1,172 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+
+import 'diagram_step.dart';
+
+const String add_event = 'add_event';
+const String arrow_menu = 'arrow_menu';
+const String close_menu = 'close_menu';
+const String ellipsis_search = 'ellipsis_search';
+const String event_add = 'event_add';
+const String home_menu = 'home_menu';
+const String list_view = 'list_view';
+const String menu_arrow = 'menu_arrow';
+const String menu_close = 'menu_close';
+const String menu_home = 'menu_home';
+const String pause_play = 'pause_play';
+const String play_pause = 'play_pause';
+const String search_ellipsis = 'search_ellipsis';
+const String view_list = 'view_list';
+
+const Duration _kAnimationDuration = Duration(seconds: 1);
+
+/// A base class for AnimatedIcons diagrams.
+abstract class AnimatedIconsDiagram<T> extends StatefulWidget
+    with DiagramMetadata {
+  const AnimatedIconsDiagram({super.key});
+
+  /// The animation curve for the animated icons.
+  Curve get curve;
+  Animation<T> buildAnimation(AnimationController controller);
+  Widget buildTransition(BuildContext context, Animation<T> animation);
+
+  @override
+  Duration? get duration => _kAnimationDuration;
+
+  @override
+  AnimatedIconsDiagramState<T> createState() => AnimatedIconsDiagramState<T>();
+}
+
+class AnimatedIconsDiagramState<T> extends State<AnimatedIconsDiagram<T>>
+    with TickerProviderStateMixin<AnimatedIconsDiagram<T>> {
+  late Animation<T> animation;
+  late AnimationController _controller;
+
+  @override
+  void initState() {
+    super.initState();
+    _controller = AnimationController(
+      duration: _kAnimationDuration,
+      vsync: this,
+    )..addListener(() {
+        setState(() {
+          // The animation controller is changing the animation value, so we
+          // need to redraw.
+        });
+      });
```

Comment:
`AnimatedBuilder` is a cleaner way of rebuilding subtrees when a Listenable updates